controllers/mysql_combined_test.go

```diff
@@ -197,3 +199,81 @@ func RunMySQLUserHappyPath(ctx context.Context, t *testing.T, mySQLServerName st
 		return true
 	}, tc.timeout, tc.retry, "waiting for DB user to be updated")
 }
+
+func TestMySQLUserSuppliedPassword(t *testing.T) {
+	t.Parallel()
+	defer PanicRecover(t)
+	ctx := context.Background()
+	assert := require.New(t)
+
+	// Add any setup steps that needs to be executed before each test
+	rgLocation := "westus2"
+	rgName := tc.resourceGroupName
+	mySQLServerName := GenerateTestResourceNameWithRandom("mysql-srv", 10)
+	mySQLReplicaName := GenerateTestResourceNameWithRandom("mysql-rep", 10)
+
+	adminSecretName := GenerateTestResourceNameWithRandom("mysqlsecret", 10)
+	adminUsername := helpers.GenerateRandomUsername(10)
+	adminPassword := helpers.NewPassword()
+	// Create the secret
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: adminSecretName,
+			Namespace: "default",
+		},
+		StringData: map[string]string{
+			"username": adminUsername,
+			"password": adminPassword,
+		},
+	}
+	err := tc.k8sClient.Create(ctx, secret)
+	assert.NoError(err)
+
+	// Create the mySQLServer object and expect the Reconcile to be created
+	mySQLServerInstance := v1alpha2.NewDefaultMySQLServer(mySQLServerName, rgName, rgLocation)
+	mySQLServerInstance.Spec.AdminSecret = adminSecretName
+	RequireInstance(ctx, t, tc, mySQLServerInstance)
+
+	// Create a mySQL replica
+	mySQLReplicaInstance := v1alpha2.NewReplicaMySQLServer(mySQLReplicaName, rgName, rgLocation, mySQLServerInstance.Status.ResourceId)
+	mySQLReplicaInstance.Spec.StorageProfile = nil
+	EnsureInstance(ctx, t, tc, mySQLReplicaInstance)
```

Comment:
The replica has two "modes":
1. The old mode, where it reads the username and password from the primary secret created by ASO.
2. The new mode, where if _on the replica_ the user specifies `AdminSecret`, it instead gets the secret directly from that location. It still reads the primary secret created by ASO for all of the other fields it needs (URLs mostly).

My intent here was to test mode 1, but actually I can do that faster by just asserting on the shape of the generated secret after creating the primary server. As long as the generated secret looks the same as it does when not using `AdminSecret` on the primary, the replica should work.

For mode 2, I didn't write a test for it. The additional code is pretty trivial (a single `if`) and as you say, additional creations of replicas are burdensome because of how long they take to make.

I'll change this to just assert the shape of the secret from the primary is as expected and drop the replica creation part.

Reply:
Actually you're right, I can just remove this entirely because the MySQLUser test already requires the username/password so this is already being verified by that.